README.md
# PHP Transmission API
[![Latest Stable Version](https://poser.pugx.org/mostertb/transmission-php/v/stable)](https://packagist.org/packages/mostertb/transmission-php)
[![Total Downloads](https://poser.pugx.org/mostertb/transmission-php/downloads)](https://packagist.org/packages/mostertb/transmission-php)
[![License](https://poser.pugx.org/mostertb/transmission-php/license)](https://packagist.org/packages/mostertb/transmission-php)
[![Build Status](https://travis-ci.com/mostertb/transmission-php.svg?branch=master)](https://travis-ci.com/mostertb/transmission-php)

This repo contains a fork of the un-maintained [kleiram/transmission-php](https://github.com/kleiram/transmission-php)
PHP Transmission RPC client library. This fork provides support for more recent versions of PHP and functionality in newer
versions of Transmission as well as fixes to upstream issues. See the [CHANGLOG](CHANGELOG) for a full list of changes 

The library provides an interface to the [Transmission](http://transmissionbt.com) bit-torrent downloader. It provides 
means to get and remove torrents from the downloader as well as adding new torrents to the download queue.

## Installation

Installation is easy using [Composer](https://getcomposer.org):

```bash
composer require mostertb/transmission-php
```

## Usage

Using the library is as easy as installing it:

```php
<?php
use Transmission\Transmission;

$transmission = new Transmission();

// Getting all the torrents currently in the download queue
$torrents = $transmission->all();

// Getting a specific torrent from the download queue
$torrent = $transmission->get(1);

// (you can also get a torrent by the hash of the torrent)
$torrent = $transmission->get(/* torrent hash */);

// Adding a torrent to the download queue
$torrent = $transmission->add(/* path to torrent */);

// Removing a torrent from the download queue
$torrent = $transmission->get(1);
$transmission->remove($torrent);

// Or if you want to delete all local data too
$transmission->remove($torrent, true);

// You can also get the Trackers that the torrent currently uses
// These are instances of the Transmission\Model\Tracker class
$trackers = $torrent->getTrackers();

// You can also get the Trackers statistics and info that the torrent currently has
// These are instances of the Transmission\Model\trackerStats class
$trackerStats = $torrent->getTrackerStats();

// To get the start date/time of the torrent in UNIX Timestamp format
$startTime = $torrent -> getStartDate();

// To get the number of peers connected
$connectedPeers = $torrent -> getPeersConnected();

// Getting the files downloaded by the torrent are available too
// These are instances of Transmission\Model\File
$files = $torrent->getFiles();

// You can start, stop, verify the torrent and ask the tracker for
// more peers to connect to
$transmission->stop($torrent);
$transmission->start($torrent);
$transmission->start($torrent, true); // Pass true if you want to start the torrent immediatly
$transmission->verify($torrent);
$transmission->reannounce($torrent);
```

To find out which information is contained by the torrent, check
[`Transmission\Model\Torrent`](https://github.com/kleiram/transmission-php/tree/master/lib/Transmission/Model/Torrent.php).

By default, the library will try to connect to `localhost:9091`. If you want to
connect to another host or post you can pass those to the constructor of the
`Transmission` class. 

```php
<?php
use Transmission\Transmission;

$transmission = new Transmission('example.com', 33);

$torrents = $transmission->all();
$torrent  = $transmission->get(1);
$torrent  = $transmission->add(/* path to torrent */);

// When you already have a torrent, you don't have to pass the client again
$torrent->delete();
```
The constructor will also allow you to configure a non-standard RPC URL with its third parameter and
a request timeout (in seconds) with its fourth  parameter. All parameters are optional:
```php
<?php
use Transmission\Transmission;

$transmission = new Transmission(null, null, '/transmission/some-custom-url/rpc', 10);
```

It is also possible to pass the torrent data directly instead of using a file
but the metadata must be base64-encoded:

```php
<?php
$torrent = $transmission->add(/* base64-encoded metainfo */, true);
```

If the Transmission server is secured with a username and password you can
authenticate using the `Client` class:

```php
<?php
use Transmission\Client;
use Transmission\Transmission;

$client = new Client(); // Can take the same optional parameters for host, port, url and timeout as the Transmission class
$client->authenticate('username', 'password');
$transmission = new Transmission();
$transmission->setClient($client);
```

Additionally, you can control the actual Transmission setting. This means
you can modify the global download limit or change the download directory:

```php
<?php
use Transmission\Transmission;

$transmission = new Transmission();
$session = $transmission->getSession();

$session->setDownloadDir('/home/foo/downloads/complete');
$session->setIncompleteDir('/home/foo/downloads/incomplete');
$session->setIncompleteDirEnabled(true);
$session->save();
```

## Testing

Testing is done using [PHPUnit](https://github.com/sebastianbergmann/phpunit). Assuming your `composer install` was done
including development dependencies, you can run the version of PHPUnit supported by this project straight from the `vendor`
directory:

```bash
php vendor/bin/phpunit
```

## License

This library is licensed under the BSD 2-clause license. Refer to the [LICENSE](LICENSE) file


CHANGELOG
Version     Changes

# Upstream Changes
## https://github.com/kleiram/transmission-php
0.1.0       - Initial release

0.2.0       - Rewrote the entire public API

0.3.0       - Added support for authentication

0.4.0       - The library now requires at least PHP 5.3.2
            - Added support for getting files downloaded by torrent
            - Added support for getting trackers used by a torrent
            - Added support for getting peers connected to
            - The torrent now contains:
                * Whether it is finished
                * The up- and download rate (in bytes/s)
                * The size of the download (when completed)
                * The ETA of the download
                * The percentage of the download completed
            - Made the authentication more flexible
            - The client now sends an User-Agent header with each request
            - Added support for starting, stopping, veryfing and
              requesting a reannounce of torrents

0.5.0       - Fix a bug in the authentication/authorization mechanism
            - A whole lot of other stuff including management of the
              Transmission session (setting global download speed limit
              and toggling the speed limit among others).

0.5.1       - Minor fix to test suite

0.5.2       - Minor fix to test suite

1.0.0       - Added support for custom RPC API path
            - Adding a duplicate torrent no longer throws an exception
            - Fix removal of local data when removing a torrent
            - Documentation updates

# Fork Changes
1.1.0       - Upgraded library and dependencies to support PHP 7. Library no requires at least PHP 7.2
            - Upgraded Test Suite to support PHPUnit 8
            - Added support for configuring HTTP timeouts between the library and Transmission client
            - Added support for all torrent stats cached in the transmission tr_stat struct as of Transmission 2.94 (22 new fields on the Torrent class)
            - Fixed loose version constraint on kriswallsmith/buzz causing installs with versions > 0.16.1 to fail
            - Saving a session that is not backed by a Transmission Client now throws an exception rather than failing silently



tests/Transmission/Tests/ClientTest.php
<?php
namespace Transmission\Tests;

use Transmission\Client;

class ClientTest extends \PHPUnit\Framework\TestCase
{
    protected $client;

    /**
     * @test
     */
    public function shouldHaveDefaultHost()
    {
        $this->assertEquals('localhost', $this->getClient()->getHost());
    }

    /**
     * @test
     */
    public function shouldHaveDefaultPort()
    {
        $this->assertEquals(9091, $this->getClient()->getPort());
    }

    /**
     * @test
     */
    public function shouldHaveNoTokenOnInstantiation()
    {
        $this->assertEmpty($this->getClient()->getToken());
    }

    /**
     * @test
     */
    public function shouldHaveDefaultClient()
    {
        $this->assertInstanceOf('Buzz\Client\Curl', $this->getClient()->getClient());
    }

    /**
     * @test
     */
    public function shouldGenerateDefaultUrl()
    {
        $this->assertEquals('http://localhost:9091', $this->getClient()->getUrl());
    }

    /**
     * @test
     */
    public function shouldMakeApiCall()
    {
        $test   = $this;
        $client = $this->getMockBuilder('Buzz\Client\Curl')->getMock();
        $client->expects($this->once())
            ->method('send')
            ->with(
                $this->isInstanceOf('Buzz\Message\Request'),
                $this->isInstanceOf('Buzz\Message\Response')
            )
            ->will($this->returnCallback(function ($request, $response) use ($test) {
                $test->assertEquals('POST', $request->getMethod());
                $test->assertEquals('/transmission/rpc', $request->getResource());
                $test->assertEquals('http://localhost:9091', $request->getHost());
                $test->assertEmpty($request->getHeader('X-Transmission-Session-Id'));
                $test->assertEquals('{"method":"foo","arguments":{"bar":"baz"}}', $request->getContent());

                $response->addHeader('HTTP/1.1 200 OK');
                $response->addHeader('Content-Type: application/json');
                $response->addHeader('X-Transmission-Session-Id: foo');
                $response->setContent('{"foo":"bar"}');
            }));

        $this->getClient()->setClient($client);
        $response = $this->getClient()->call('foo', array('bar' => 'baz'));

        $this->assertInstanceOf('stdClass', $response);
        $this->assertObjectHasAttribute('foo', $response);
        $this->assertEquals('bar', $response->foo);
    }

    /**
     * @test
     */
    public function shouldAuthenticate()
    {
        $test   = $this;
        $client = $this->getMockBuilder('Buzz\Client\Curl')->getMock();
        $client->expects($this->once())
            ->method('send')
            ->with(
                $this->isInstanceOf('Buzz\Message\Request'),
                $this->isInstanceOf('Buzz\Message\Response')
            )
            ->will($this->returnCallback(function ($request, $response) use ($test) {
                $test->assertEquals('POST', $request->getMethod());
                $test->assertEquals('/transmission/rpc', $request->getResource());
                $test->assertEquals('http://localhost:9091', $request->getHost());
                $test->assertEmpty($request->getHeader('X-Transmission-Session-Id'));
                $test->assertEquals('Basic '. base64_encode('foo:bar'), $request->getHeader('Authorization'));
                $test->assertEquals('{"method":"foo","arguments":{"bar":"baz"}}', $request->getContent());

                $response->addHeader('HTTP/1.1 200 OK');
                $response->addHeader('Content-Type: application/json');
                $response->addHeader('X-Transmission-Session-Id: foo');
                $response->setContent('{"foo":"bar"}');
            }));

        $this->getClient()->authenticate('foo', 'bar');
        $this->getClient()->setClient($client);
        $response = $this->getClient()->call('foo', array('bar' => 'baz'));

        $this->assertInstanceOf('stdClass', $response);
        $this->assertObjectHasAttribute('foo', $response);
        $this->assertEquals('bar', $response->foo);
    }

    /**
     * @test
     */
    public function shouldThrowExceptionOnExceptionDuringApiCall()
    {
        $client = $this->getMockBuilder('Buzz\Client\Curl')->getMock();
        $client->expects($this->once())
            ->method('send')
            ->will($this->throwException(new \Exception()));

        $this->getClient()->setClient($client);
        $this->expectException('RuntimeException');
        $this->getClient()->call('foo', array());
    }

    /**
     * @test
     */
    public function shouldThrowExceptionOnUnexpectedStatusCode()
    {
        $client = $this->getMockBuilder('Buzz\Client\Curl')->getMock();
        $client->expects($this->once())
            ->method('send')
            ->will($this->returnCallback(function ($request, $response) {
                $response->addHeader('HTTP/1.1 500 Internal Server Error');
            }));

        $this->getClient()->setClient($client);
        $this->expectException('RuntimeException');
        $this->getClient()->call('foo', array());
    }

    /**
     * @test
     */
    public function shouldThrowExceptionOnAccessDenied()
    {
        $client = $this->getMockBuilder('Buzz\Client\Curl')->getMock();
        $client->expects($this->once())
            ->method('send')
            ->will($this->returnCallback(function ($request, $response) {
                $response->addHeader('HTTP/1.1 401 Access Denied');
            }));

        $this->getClient()->setClient($client);
        $this->expectException('RuntimeException');
        $this->getClient()->call('foo', array());
    }

    /**
     * @test
     */
    public function shouldHandle409ResponseWhenMakingAnApiCall()
    {
        $test   = $this;
        $client = $this->getMockBuilder('Buzz\Client\Curl')->getMock();
        $client->expects($this->at(0))
            ->method('send')
            ->will($this->returnCallback(function ($request, $response) use ($test) {
                $test->assertEmpty($request->getHeader('X-Transmission-Session-Id'));

                $response->addHeader('HTTP/1.1 409 Conflict');
                $response->addHeader('X-Transmission-Session-Id: foo');
            }));

        $client->expects($this->at(1))
            ->method('send')
            ->will($this->returnCallback(function ($request, $response) {
                $response->addHeader('HTTP/1.1 200 OK');
                $response->addHeader('Content-Type: application/json');
                $response->addHeader('X-Transmission-Session-Id: foo');
                $response->setContent('{"foo":"bar"}');
            }));

        $this->getClient()->setClient($client);
        $response = $this->getClient()->call('foo', array());

        $this->assertEquals('foo', $this->getClient()->getToken());
        $this->assertInstanceOf('stdClass', $response);
        $this->assertObjectHasAttribute('foo', $response);
        $this->assertEquals('bar', $response->foo);
    }

    public function setup(): void
    {
        $this->client = new Client();
    }

    private function getClient()
    {
        return $this->client;
    }
}


tests/Transmission/Tests/Model/FileTest.php
<?php
namespace Transmission\Tests\Model;

use Transmission\Model\File;
use Transmission\Util\PropertyMapper;


class FileTest extends \PHPUnit\Framework\TestCase
{
    protected $file;

    /**
     * @test
     */
    public function shouldImplementModelInterface()
    {
        $this->assertInstanceOf('Transmission\Model\ModelInterface', $this->getFile());
    }

    /**
     * @test
     */
    public function shouldHaveNonEmptyMapping()
    {
        $this->assertNotEmpty($this->getFile()->getMapping());
    }

    /**
     * @test
     */
    public function shouldBeCreatedFromMapping()
    {
        $source = (object) array(
            'name' => 'foo',
            'length' => 100,
            'bytesCompleted' => 10
        );

        PropertyMapper::map($this->getFile(), $source);

        $this->assertEquals('foo', $this->getFile()->getName());
        $this->assertEquals(100, $this->getFile()->getSize());
        $this->assertEquals(10, $this->getFile()->getCompleted());
        $this->assertFalse($this->getFile()->isDone());
    }

    /**
     * @test
     */
    public function shouldConvertToString()
    {
        $this->getFile()->setName('foo');

        $this->assertIsString((string) $this->getFile());
        $this->assertEquals('foo', (string) $this->getFile());
    }

    public function setup(): void
    {
        $this->file = new File();
    }

    private function getFile()
    {
        return $this->file;
    }
}



tests/Transmission/Tests/Model/PeerTest.php
<?php
namespace Transmission\Tests\Model;

use Transmission\Model\Peer;
use Transmission\Util\PropertyMapper;

class PeerTest extends \PHPUnit\Framework\TestCase
{
    protected $peer;

    /**
     * @test
     */
    public function shouldImplementModelInterface()
    {
        $this->assertInstanceOf('Transmission\Model\ModelInterface', $this->getPeer());
    }

    /**
     * @test
     */
    public function shouldHaveNonEmptyMapping()
    {
        $this->assertNotEmpty($this->getPeer()->getMapping());
    }

    /**
     * @test
     */
    public function shouldBeCreatedFromMapping()
    {
        $source = (object) array(
            'address' => 'foo',
            'clientName' => 'foo',
            'clientIsChoked' => false,
            'clientIsInterested' => true,
            'flagStr' => 'foo',
            'isDownloadingFrom' => false,
            'isEncrypted' => true,
            'isIncoming' => false,
            'isUploadingTo' => true,
            'isUTP' => false,
            'peerIsChoked' => true,
            'peerIsInterested' => false,
            'port' => 3000,
            'progress' => 10.5,
            'rateToClient' => 1000,
            'rateFromClient' => 10000
        );

        PropertyMapper::map($this->getPeer(), $source);

        $this->assertEquals('foo', $this->getPeer()->getAddress());
        $this->assertEquals('foo', $this->getPeer()->getClientName());
        $this->assertFalse($this->getPeer()->isClientChoked());
        $this->assertTrue($this->getPeer()->isClientInterested());
        $this->assertFalse($this->getPeer()->isDownloading());
        $this->assertTrue($this->getPeer()->isEncrypted());
        $this->assertFalse($this->getPeer()->isIncoming());
        $this->assertTrue($this->getPeer()->isUploading());
        $this->assertFalse($this->getPeer()->isUtp());
        $this->assertTrue($this->getPeer()->isPeerChoked());
        $this->assertFalse($this->getPeer()->isPeerInterested());
        $this->assertEquals(3000, $this->getPeer()->getPort());
        $this->assertEquals(10.5, $this->getPeer()->getProgress());
        $this->assertEquals(1000, $this->getPeer()->getUploadRate());
        $this->assertEquals(10000, $this->getPeer()->getDownloadRate());
    }

    public function setup(): void
    {
        $this->peer = new Peer();
    }

    public function getPeer()
    {
        return $this->peer;
    }
}


lib/Transmission/Transmission.php
<?php
namespace Transmission;

use Transmission\Model\Torrent;
use Transmission\Model\Session;
use Transmission\Model\FreeSpace;
use Transmission\Model\Stats\Session as SessionStats;
use Transmission\Util\PropertyMapper;
use Transmission\Util\ResponseValidator;

/**
 * @author Ramon Kleiss <ramon@cubilon.nl>
 */
class Transmission
{
    /**
     * @var Client
     */
    protected $client;

    /**
     * @var ResponseValidator
     */
    protected $validator;

    /**
     * @var PropertyMapper
     */
    protected $mapper;

    /**
     * Constructor
     *
     * @param string  $host The hostname or IP of the Transmission server
     * @param integer $port The port the Transmission server is listening on
     * @param string  $path The path to Transmission server rpc api
     * @param integer $timeout Number of seconds after which to fail requests
     */
    public function __construct($host = null, $port = null, $path = null, $timeout = null)
    {
        $this->setClient(new Client($host, $port, $path, $timeout));
        $this->setMapper(new PropertyMapper());
        $this->setValidator(new ResponseValidator());
    }

    /**
     * Get all the torrents in the download queue
     *
     * @return Torrent[]
     */
    public function all()
    {
        $client   = $this->getClient();
        $mapper   = $this->getMapper();
        $response = $this->getClient()->call(
            'torrent-get',
            array('fields' => array_keys(Torrent::getMapping()))
        );

        return array_map(function ($data) use ($mapper, $client) {
            return $mapper->map(
                new Torrent($client),
                $data
            );
        }, $this->getValidator()->validate('torrent-get', $response));
    }

    /**
     * Get a specific torrent from the download queue
     *
     * @param  integer                    $id
     * @return Torrent
     * @throws \RuntimeException
     */
    public function get($id)
    {
        $client   = $this->getClient();
        $mapper   = $this->getMapper();
        $response = $this->getClient()->call('torrent-get', array(
            'fields' => array_keys(Torrent::getMapping()),
            'ids'    => array($id)
        ));

        $torrent = array_reduce(
            $this->getValidator()->validate('torrent-get', $response),
            function ($torrent, $data) use ($mapper, $client) {
                return $torrent ? $torrent : $mapper->map(new Torrent($client), $data);
            });

        if (!$torrent instanceof Torrent) {
            throw new \RuntimeException(sprintf("Torrent with ID %s not found", $id));
        }

        return $torrent;
    }

    /**
     * Get the Transmission session
     *
     * @return Session
     */
    public function getSession()
    {
        $response = $this->getClient()->call(
            'session-get',
            array()
        );

        return $this->getMapper()->map(
            new Session($this->getClient()),
            $this->getValidator()->validate('session-get', $response)
        );
    }

    /**
     * @return SessionStats
     */
    public function getSessionStats()
    {
        $response = $this->getClient()->call(
            'session-stats',
            array()
        );

        return $this->getMapper()->map(
            new SessionStats(),
            $this->getValidator()->validate('session-stats', $response)
        );
    }

    /**
     * Get Free space
     * @param  string $path
     * @return FreeSpace
     */
    public function getFreeSpace($path=null)
    {
        if (!$path) {
            $path = $this->getSession()->getDownloadDir();
        }
        $response = $this->getClient()->call(
            'free-space',
            array('path'=>$path)
        );

        return $this->getMapper()->map(
            new FreeSpace(),
            $this->getValidator()->validate('free-space', $response)
        );
    }

    /**
     * Add a torrent to the download queue
     *
     * @param  string   $torrent
     * @param  boolean  $metainfo
     * @param  string   $savepath
     * @return Torrent
     */
    public function add($torrent, $metainfo = false, $savepath = null)
    {
        $parameters = array($metainfo ? 'metainfo' : 'filename' => $torrent);

        if ($savepath !== null) {
            $parameters['download-dir'] = (string) $savepath;
        }

        $response = $this->getClient()->call(
            'torrent-add',
            $parameters
        );

        return $this->getMapper()->map(
            new Torrent($this->getClient()),
            $this->getValidator()->validate('torrent-add', $response)
        );
    }

    /**
     * Start the download of a torrent
     *
     * @param Torrent $torrent
     * @param bool    $now
     */
    public function start(Torrent $torrent, $now = false)
    {
        $this->getClient()->call(
            $now ? 'torrent-start-now' : 'torrent-start',
            array('ids' => array($torrent->getId()))
        );
    }

    /**
     * Stop the download of a torrent
     *
     * @param Torrent $torrent
     */
    public function stop(Torrent $torrent)
    {
        $this->getClient()->call(
            'torrent-stop',
            array('ids' => array($torrent->getId()))
        );
    }

    /**
     * Verify the download of a torrent
     *
     * @param Torrent $torrent
     */
    public function verify(Torrent $torrent)
    {
        $this->getClient()->call(
            'torrent-verify',
            array('ids' => array($torrent->getId()))
        );
    }

    /**
     * Request a reannounce of a torrent
     *
     * @param Torrent $torrent
     */
    public function reannounce(Torrent $torrent)
    {
        $this->getClient()->call(
            'torrent-reannounce',
            array('ids' => array($torrent->getId()))
        );
    }

    /**
     * Remove a torrent from the download queue
     *
     * @param Torrent $torrent
     */
    public function remove(Torrent $torrent, $localData = false)
    {
        $arguments = array('ids' => array($torrent->getId()));

        if ($localData) {
            $arguments['delete-local-data'] = true;
        }

        $this->getClient()->call('torrent-remove', $arguments);
    }

    /**
     * Set the client used to connect to Transmission
     *
     * @param Client $client
     */
    public function setClient(Client $client)
    {
        $this->client = $client;
    }

    /**
     * Get the client used to connect to Transmission
     *
     * @return Client
     */
    public function getClient()
    {
        return $this->client;
    }

    /**
     * Set the hostname of the Transmission server
     *
     * @param string $host
     */
    public function setHost($host)
    {
        $this->getClient()->setHost($host);
    }

    /**
     * Get the hostname of the Transmission server
     *
     * @return string
     */
    public function getHost()
    {
        return $this->getClient()->getHost();
    }

    /**
     * Set the port the Transmission server is listening on
     *
     * @param integer $port
     */
    public function setPort($port)
    {
        $this->getClient()->setPort($port);
    }

    /**
     * Get the port the Transmission server is listening on
     *
     * @return integer
     */
    public function getPort()
    {
        return $this->getClient()->getPort();
    }

    /**
     * Set the mapper used to map responses from Transmission to models
     *
     * @param PropertyMapper $mapper
     */
    public function setMapper(PropertyMapper $mapper)
    {
        $this->mapper = $mapper;
    }

    /**
     * Get the mapper used to map responses from Transmission to models
     *
     * @return PropertyMapper
     */
    public function getMapper()
    {
        return $this->mapper;
    }

    /**
     * Set the validator used to validate Transmission responses
     *
     * @param ResponseValidator $validator
     */
    public function setValidator(ResponseValidator $validator)
    {
        $this->validator = $validator;
    }

    /**
     * Get the validator used to validate Transmission responses
     *
     * @return ResponseValidator
     */
    public function getValidator()
    {
        return $this->validator;
    }
}


lib/Transmission/Client.php
<?php
namespace Transmission;

use Buzz\Message\Request;
use Buzz\Message\Response;
use Buzz\Client\Curl;
use Buzz\Client\ClientInterface;

/**
 * The Client class is used to make API calls to the Transmission server
 *
 * @author Ramon Kleiss <ramon@cubilon.nl>
 */
class Client
{
    /**
     * @var string
     */
    const DEFAULT_HOST = 'localhost';

    /**
     * @var integer
     */
    const DEFAULT_PORT = 9091;

    /**
     * @var string
     */
    const DEFAULT_PATH = '/transmission/rpc';

    /**
     * @var string
     */
    const TOKEN_HEADER = 'X-Transmission-Session-Id';

    /**
     * @var string
     */
    protected $host = self::DEFAULT_HOST;

    /**
     * @var integer
     */
    protected $port = self::DEFAULT_PORT;

    /**
     * @var string
     */
    protected $path = self::DEFAULT_PATH;

    /**
     * @var string
     */
    protected $token;

    /**
     * @var ClientInterface
     */
    protected $client;

    /**
     * @var string
     */
    protected $auth;

    /**
     * Constructor
     *
     * @param string  $host The hostname or IP of the Transmission server
     * @param integer $port The port the Transmission server is listening on
     * @param string  $path The path to Transmission server rpc api
     * @param integer $timeout Number of seconds after which to fail requests
     */
    public function __construct($host = null, $port = null, $path = null, $timeout = null)
    {
        $this->token    = null;
        $this->client   = new Curl();

        if(!is_null($timeout)){
            $this->client->setTimeout($timeout);
        }

        if ($host) $this->setHost($host);
        if ($port) $this->setPort($port);
        if ($path) $this->setPath($path);
    }

    /**
     * Authenticate against the Transmission server
     *
     * @param string $username
     * @param string $password
     */
    public function authenticate($username, $password)
    {
        $this->auth = base64_encode($username .':'. $password);
    }

    /**
     * Make an API call
     *
     * @param  string           $method
     * @param  array            $arguments
     * @return \stdClass
     * @throws \RuntimeException
     */
    public function call($method, array $arguments)
    {
        [$request, $response] = $this->compose($method, $arguments);

        try {
            $this->getClient()->send($request, $response);
        } catch (\Exception $e) {
            throw new \RuntimeException(
                'Could not connect to Transmission',
                0,
                $e
            );
        }

        return $this->validateResponse($response, $method, $arguments);
    }

    /**
     * Get the URL used to connect to Transmission
     *
     * @return string
     */
    public function getUrl()
    {
        return sprintf(
            'http://%s:%d',
            $this->getHost(),
            $this->getPort()
        );
    }

    /**
     * Set the hostname of the Transmission server
     *
     * @param string $host
     */
    public function setHost($host)
    {
        $this->host = (string) $host;
    }

    /**
     * Get the hostname of the Transmission server
     *
     * @return string
     */
    public function getHost()
    {
        return $this->host;
    }

    /**
     * Set the port the Transmission server is listening on
     *
     * @param integer $port
     */
    public function setPort($port)
    {
        $this->port = (integer) $port;
    }

    /**
     * Get the port the Transmission server is listening on
     *
     * @return integer
     */
    public function getPort()
    {
        return $this->port;
    }

    /**
     * Set the path to Transmission server rpc api
     *
     * @param string $path
     */
    public function setPath($path)
    {
        return $this->path = (string) $path;
    }

    /**
     * Get the path to Transmission server rpc api
     */
    public function getPath()
    {
        return $this->path;
    }

    /**
     * Set the CSRF-token of the Transmission client
     *
     * @param string $token
     */
    public function setToken($token)
    {
        $this->token = (string) $token;
    }

    /**
     * Get the CSRF-token for the Transmission client
     *
     * @return string
     */
    public function getToken()
    {
        return $this->token;
    }

    /**
     * Set the Buzz client used to connect to Transmission
     *
     * @param ClientInterface $client
     */
    public function setClient(ClientInterface $client)
    {
        $this->client = $client;
    }

    /**
     * Get the Buzz client used to connect to Transmission
     *
     * @return ClientInterface
     */
    public function getClient()
    {
        return $this->client;
    }

    /**
     * @param string $method
     * @param array $arguments
     * @return array|Request[]|Response[]
     */
    protected function compose($method, $arguments)
    {
        $request = new Request('POST', $this->getPath(), $this->getUrl());
        $request->addHeader(sprintf('%s: %s', self::TOKEN_HEADER, $this->getToken()));
        $request->setContent(json_encode(array(
            'method'    => $method,
            'arguments' => $arguments
        )));

        if (is_string($this->auth)) {
            $request->addHeader(sprintf('Authorization: Basic %s', $this->auth));
        }

        return array($request, new Response());
    }

    /**
     * @param  Response $response
     * @param  string                $method
     * @param  array                 $arguments
     * @return \stdClass
     * @throws \RuntimeException
     */
    protected function validateResponse($response, $method, $arguments)
    {
        if (!in_array($response->getStatusCode(), array(200, 401, 409))) {
            throw new \RuntimeException('Unexpected response received from Transmission');
        }

        if ($response->getStatusCode() == 401) {
            throw new \RuntimeException('Access to Transmission requires authentication');
        }

        if ($response->getStatusCode() == 409) {
            $this->setToken($response->getHeader(self::TOKEN_HEADER));

            return $this->call($method, $arguments);
        }

        return json_decode($response->getContent());
    }
}
